### ISSUE TITLE
False positive unnecessary-lambda with sum of datetime and timedelta.

### ISSUE THREAD

**Issue by bbcVld** [PyCQA/pylint#3808]
### Steps to reproduce
1. Create next files:
    test.py

```
from datetime import datetime, timedelta as td


see_you = {
    'now': lambda: datetime.utcnow().isoformat(),  # it's ok
    'next_time': lambda: (datetime.utcnow() + td(hours=6)).isoformat(),  # false positive
}
```
2. Run pylint test.py

### Current behavior
```
************* Module test
test.py:6:17: W0108: Lambda may not be necessary (unnecessary-lambda)

------------------------------------------------------------------
Your code has been rated at 5.00/10 (previous run: 4.00/10, +1.00)
```

### Expected behavior
```
Your code has been rated at 10.00/10
```

### pylint --version output
```
pylint 2.5.3
astroid 2.4.2
Python 3.7.6 (default, Dec 24 2019, 16:49:40i[GCC 5.5.0 20171010 (ROSA)]
```
### Additional info

I tried same thing with random.choice (`lambda: (choice(['aa', 'bb', 'cc']) + '.com').lower()`) and got same behavior. So I think it's a class of problems.

**Comment by dbaty** [comment]
As far as I understand, pylint tries to determine if the content of the lambda is callable and if it takes the same number of arguments as the callable. For example, with the following code:

```python
def func():
    return 2

handler = lambda: func()
```

pylint would issue a warning here because the last line should be rewritten like this:

```python
handler = func
```

Obviously, your code could not be rewritten as `{'next_time': (datetime.utcnow() + td(hours=6)).isoformat}` (without the last two parentheses) because `datetime.utcnow()` does not always return the same value (which is why you put it in a lambda). But I don't think that pylint can detect that. Hence the warning.

I was more suprised by the fact that `lambda: datetime.utcnow().isoformat()` does *not* report any warning. I found out that this case is handled here: https://github.com/PyCQA/pylint/blob/8112aae184f73bb780c7918821f0cac7fef7c74a/pylint/checkers/base.py#L1266-L1271

All in all, I don't think that this is a false positive.

**Comment by AWhetter** [comment]
I agree with @dbaty.

pylint doesn't know that `datetime.utcnow()` returns different things depending on the state at run time. The check assumes that all the functions being called in the lambda are pure (it's return value depends on nothing other than it's input arguments). Depending on your style of programming, most of the time that's not the case.
I think it makes sense to add an option to allow you to change the behaviour of this checker so that if more than one function call is chained it won't raise the warning, and to default it to this new behaviour.

**Comment by bbcVld** [comment]
It is possible to add _runtime-state-function_ option (like _ignored-classes_ and _ignored-modules_) with list of functions and methods that are not pure function? So it don't have to assume that all functions are pure. I think it'll be more helpful.
